xbmc/music/MusicDatabase.cpp

```diff
@@ -1011,60 +1018,102 @@ bool CMusicDatabase::GetAlbum(int idAlbum, CAlbum& album, bool getSongs /* = tru
     }
 
     int albumArtistOffset = album_enumCount;
-    int songOffset = albumArtistOffset + artistCredit_enumCount;
-    int songArtistOffset = songOffset + song_enumCount;
-    int infoSongOffset = songArtistOffset + artistCredit_enumCount;
-
+    
     std::set<int> artistcredits;
-    std::set<int> songs;
-    std::set<std::pair<int, int> > songartistcredits;
-    std::set<int> infosongs;
+    std::set<int> artistroles;
     album = GetAlbumFromDataset(m_pDS.get()->get_sql_record(), 0, true); // true to grab and parse the imageURL
     while (!m_pDS->eof())
     {
       const dbiplus::sql_record* const record = m_pDS->get_sql_record();
 
-      // Because rows repeat in the joined query (cartesian join) we may see each
-      // entity (album artist, song, song artist) multiple times in the result set.
-      // Since there should never be a song with the same artist twice, or an album
-      // with the same song (by id) listed twice, we key on the entity ID and only
-      // create an entity for the first occurence of each entity in the data set.
+      // Because album artists may have multiple roles there can be more than one
+      // row in the result set for each artist credit.      
       int idAlbumArtist = record->at(albumArtistOffset + artistCredit_idArtist).get_asInt();
+      int idArtistRole = record->at(albumArtistOffset + artistCredit_idRole).get_asInt();
       if (artistcredits.find(idAlbumArtist) == artistcredits.end())
       {
         album.artistCredits.push_back(GetArtistCreditFromDataset(record, albumArtistOffset));
         artistcredits.insert(idAlbumArtist);
+        artistroles.clear(); //New artist so clear set of roles
+        artistroles.insert(idArtistRole);
+      }
+      else if (artistroles.find(idArtistRole) == artistroles.end())
+      {
+        album.artistCredits.back().m_Roles.push_back(idArtistRole);
+        artistroles.insert(idArtistRole);
+      }
+      m_pDS->next();
+    }
+    m_pDS->close(); // cleanup recordset data
+
+    //Get song data
+    if (getSongs)
+    {
+      sql = PrepareSQL("SELECT songview.*, songartistview.*"
+        " FROM songview "
+        " LEFT JOIN songartistview ON songview.idSong = songartistview.idSong "
```

Comment:
I can confirm for sure that there's case where albums / songs have no artists it's for singles as I use that to detect those singles since there's other problems with some returned fields to JSON.